### PR TITLE
refactor: more explicit error confirmation for http query route error.

### PR DESCRIPTION
**Pull request by youngsofun**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

also including the case that server restart (assert: always with new node_id).
we can tell it from query timeout at once.

- need to test with the clients
    - [x] go 
    - [x] java
    - [x] bendsql  
    - [x] py

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15621)

### CODE REVIEW

**Review by sundy-li — APPROVED** [review]

File: src/query/service/src/servers/http/v1/http_query_handlers.rs [hunk 1]
```diff
@@ -53,20 +53,20 @@ const HEADER_QUERY_ID: &str = "X-DATABEND-QUERY-ID";
 const HEADER_QUERY_STATE: &str = "X-DATABEND-QUERY-STATE";
 const HEADER_QUERY_PAGE_ROWS: &str = "X-DATABEND-QUERY-PAGE-ROWS";
 
-pub fn make_page_uri(query_id: &str, page_no: usize) -> String {
-    format!("/v1/query/{}/page/{}", query_id, page_no)
+pub fn make_page_uri(query_id: &str, node_id: &str, page_no: usize) -> String {
+    format!("/v1/query/{}/{}/page/{}", query_id, node_id, page_no)
```
Comment (flaneur2020) [inline src/query/service/src/servers/http/v1/http_query_handlers.rs:57]: can we pass node_id in the header?

**Review by flaneur2020 — CHANGES_REQUESTED** [review]
it need a comprehensive test over all the drivers after api route changes, and there're also few in-house query clients (like PHP) not using our drivers at all. if it's an enhancement over the error message, i guess we can consider a less invasive way to make it.

File: src/query/service/src/servers/http/v1/http_query_handlers.rs (hunk 1 above)
Comment (youngsofun) [reply src/query/service/src/servers/http/v1/http_query_handlers.rs]: done

**Comment by youngsofun** [comment]
@flaneur2020 plz review again

File: tests/suites/1_stateful/09_http_handler/09_0006_route_error.result [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# error
+## page
+{"error":{"code":"404","message":"route error: query QID SHOULD be on server XXX, but current server is NODE, which started ... ago)"}}
+## kill
+{"error":{"code":"404","message":"route error: query QID SHOULD be on server XXX, but current server is NODE, which started ... ago)"}}
+## final
+{"error":{"code":"404","message":"route error: query QID SHOULD be on server XXX, but current server is NODE, which started ... ago)"}}
```
Comment (flaneur2020) [inline tests/suites/1_stateful/09_http_handler/09_0006_route_error.result:7]: @hantmac please take a look about whether the drivers could handle this error response into a proper exception with these message displayed.

**Review by flaneur2020 — APPROVED** [review]
LGTM

File: tests/suites/1_stateful/09_http_handler/09_0006_route_error.result (hunk 2 above)
Comment (youngsofun) [reply tests/suites/1_stateful/09_http_handler/09_0006_route_error.result:7]: client cannot handle it, and it is hard to explain it in the error message, current message gives info helpful for debugging.

 we can later add a section about the routing problem in the doc, and in the error message let user to refer to the section.

**Pull request merged**